### PR TITLE
[everflow] Add IPv6 Match Tests to Everflow testbed

### PR DESCRIPTION
**Pull request by daall**
- Add tests to check IPv6 match types for Everflow
- Pivot to ptfadapter for Everflow testing
- Mark areas to refactor in future PRs

Signed-off-by: Danny Allen <daall@microsoft.com>

### Description of PR

Summary: New tests to verify IPv6 match types for Everflow.

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

#### What is the motivation for this PR?
We have functional tests for Everflow that ensure that the mirror session and policer behavior are correct, but we don't have tests that verify the different match types that need to be supported by the MIRRORV6 table. This PR addresses that gap.

#### How did you do it?
I came up with a list of rules that exercise 1) each individual qualifier in the MIRRORV6 table and 2) different combinations of qualifiers in the MIRRORV6 table. Combinations were chosen to try to replicate real-life scenarios as well as edge cases that may confuse the SAI implementation (e.g. matching a UDP packet with TCP flags).

#### How did you verify/test it?
Ran the tests against a variety of SKUs (BRCM and MLNX). Also ran the existing Everflow tests to ensure there was no regression.

```
daall@00c187053803:/var/src/sonic-mgmt/tests$ pytest everflow/test_everflow_ipv6.py --testbed=vms12-t1-lag-1 --inventory=../ansible/str --testbed_file=../ansible/testbed.csv --host-pattern=vms12-t1-lag-1 --module-path=../ansible/library -v
=============================================================================================== test session starts ================================================================================================
platform linux2 -- Python 2.7.12, pytest-4.6.5, py-1.8.1, pluggy-0.13.1 -- /usr/bin/python
cachedir: .pytest_cache
ansible: 2.8.7
rootdir: /var/src/Networking-acs-sonic-mgmt/tests, inifile: pytest.ini
plugins: forked-1.1.3, xdist-1.28.0, ansible-2.2.2, repeat-0.8.0
collected 19 items                                                                                                                                                                                                 

everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_src_ipv6_mirroring PASSED                                                                                                                      [  5%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_dst_ipv6_mirroring PASSED                                                                                                                      [ 10%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_next_header_mirroring PASSED                                                                                                                   [ 15%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_l4_src_port_mirroring PASSED                                                                                                                   [ 21%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_l4_dst_port_mirroring PASSED                                                                                                                   [ 26%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_l4_src_port_range_mirroring PASSED                                                                                                             [ 31%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_l4_dst_port_range_mirroring PASSED                                                                                                             [ 36%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_tcp_flags_mirroring PASSED                                                                                                                     [ 42%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_dscp_mirroring PASSED                                                                                                                          [ 47%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_l4_range_mirroring PASSED                                                                                                                      [ 52%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_tcp_response_mirroring PASSED                                                                                                                  [ 57%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_tcp_application_mirroring PASSED                                                                                                               [ 63%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_udp_application_mirroring PASSED                                                                                                               [ 68%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_any_protocol PASSED                                                                                                                            [ 73%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_fuzzy_packet XFAIL                                                                                                                             [ 78%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_source_subnet PASSED                                                                                                                           [ 84%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_dest_subnet PASSED                                                                                                                             [ 89%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_both_subnets PASSED                                                                                                                            [ 94%]
everflow/test_everflow_ipv6.py::TestIngressEverflowIPv6::test_fuzzy_subnets PASSED                                                                                                                           [100%]

============================================================================================= short test summary info ==============================================================================================
```

#### Any platform specific information?
N/A

#### Supported testbed topology if it's a new test case?
Same as existing Everflow tests.

### Other Notes
I will be porting the existing Everflow tests to use ptfadapter in a follow-on PR. It's much easier to use and debug than the ptf_runner implementation. With this change, we can also refactor more of the existing tests and these new tests to use a common infrastructure.

I also want to somehow parametrize/templatize the acl.json file. There is a lot of manual effort in maintaining the test cases and the acl rules file, so we should try to generate the rules file based on the set of test cases somehow.

### CODE REVIEW

**Comment by daall** [comment]
retest vsimage please

File: tests/everflow/templates/acl_rule_v6.json.j2 [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+                                    "config": {
+                                        "source-ip-address": "ffbe:0225:7c6b:a982:d48b:230e:f271:000e/128",
+                                        "destination-ip-address": "ffbe:0225:7c6b:a982:d48b:230e:f271:000f/128",
+                                        "protocol": 17,
+                                        "dscp": 16
+                                    }
+                                },
+                                "transport": {
+                                    "config": {
+                                        "source-port": "12004",
```
Comment (abdosi) [inline tests/everflow/templates/acl_rule_v6.json.j2:370]: is this negative test case ? Protocol being UDP but match on TCP Flags. What is expected behaviour here ?

File: tests/everflow/test_everflow_ipv6.py [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+        )
+
+        expected_packet["GRE"].proto = mirror_session["session_gre"]
+
+        expected_packet = Mask(expected_packet)
+        expected_packet.set_do_not_care_scapy(packet.Ether, "dst")
+        expected_packet.set_do_not_care_scapy(packet.IP, "ihl")
+        expected_packet.set_do_not_care_scapy(packet.IP, "len")
+        expected_packet.set_do_not_care_scapy(packet.IP, "flags")
+        expected_packet.set_do_not_care_scapy(packet.IP, "chksum")
```
Comment (abdosi) [inline tests/everflow/test_everflow_ipv6.py:475]: IPv6 Header does not have checksum.

File: tests/everflow/test_everflow_ipv6.py [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+            table_name = "EVERFLOWV6"
+            temporary_table = True
+            duthost.command("config acl add table {} MIRRORV6".format(table_name))
+
+        duthost.host.options["variable_manager"].extra_vars.update({"acl_table_name": table_name})
+        duthost.template(src="everflow/templates/acl_rule_v6.json.j2", dest="everflow_v6_tests/acl_rule_v6.json")
+        duthost.command("acl-loader update full everflow_v6_tests/acl_rule_v6.json --session_name={}".format(setup_mirror_session["session_name"]))
+
+        yield
+
```
Comment (abdosi) [inline tests/everflow/test_everflow_ipv6.py:572]: We don't have mirror with  policer use case for Ipv6 Traffic ?

File: tests/everflow/test_everflow_ipv6.py [hunk 4]
```diff
@@ -0,0 +1,592 @@
...
+        logging.info("received: %s", packet.Ether(received_packet).summary())
+
+        inner_packet = self._extract_mirror_payload(received_packet, len(mirror_packet))
+        logging.info("inner_packet: %s", inner_packet.summary())
+        logging.info("expected_packet: %s", mirror_packet.summary())
+        pytest_assert(Mask(inner_packet).pkt_match(mirror_packet), "Mirror payload does not match received packet")
+
+    def _get_src_port(self, setup):
+        return setup["port_index_map"][random.choice(setup["port_index_map"].keys())]
```
Comment (abdosi) [inline tests/everflow/test_everflow_ipv6.py:427]: there is difference between v4 and v6. 
For v4 we are using setup_info based on tor/spine and run test vase for both port as tor and spine.
Looks like here we execute test case only once.

File: tests/everflow/test_everflow_ipv6.py [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+                                            duthost,
+                                            test_packet)
+
+    def test_any_protocol(self, setup_info, setup_mirror_session, ptfadapter, duthost):
+        """Verify that ALL protocols are matched if no Next Header is specified."""
+        test_packet = self._base_tcp_packet(
+            ptfadapter,
+            setup_info,
+            src_ip="ffbe:0225:7c6b:a982:d48b:230e:f271:000c",
+            dst_ip="ffbe:0225:7c6b:a982:d48b:230e:f271:000d",
```
Comment (abdosi) [inline tests/everflow/test_everflow_ipv6.py:272]: will it not be TCP packet even though we did not specify Next Header .
Same for UDP below

File: tests/everflow/test_everflow_ipv6.py (hunk 3 above)
Comment (daall) [reply tests/everflow/test_everflow_ipv6.py:572]: We typically use a policer with the MIRROR_DSCP table.

File: tests/everflow/test_everflow_ipv6.py (hunk 4 above)
Comment (daall) [reply tests/everflow/test_everflow_ipv6.py:427]: We're not testing the forwarding behavior for the mirror session in this test case, so I left the tor/spine part off. Is there another reason to try both?

File: tests/everflow/test_everflow_ipv6.py (hunk 5 above)
Comment (daall) [reply tests/everflow/test_everflow_ipv6.py:272]: It will be, yeah! I think the comment is misleading. We're basically testing if we have a rule that tests src ip/dst ip/src port/dst port, then we can have any Next Header we want and the rule will still match. I'll update the comment.

File: tests/everflow/templates/acl_rule_v6.json.j2 (hunk 1 above)
Comment (daall) [reply tests/everflow/templates/acl_rule_v6.json.j2:370]: We want to see if the ASIC crashes when we install this rule. The idea being that the user could specify an arbitrary protocol number and we want to make sure that we don't crash when the protocol number conflicts with other fields in the ACL rule.

File: tests/everflow/test_everflow_ipv6.py (hunk 4 above)
Comment (abdosi) [reply tests/everflow/test_everflow_ipv6.py:427]: yeah then it is fine.

File: tests/everflow/templates/acl_rule_v6.json.j2 (hunk 1 above)
Comment (abdosi) [reply tests/everflow/templates/acl_rule_v6.json.j2:370]: ok, then will Sonic return error and there will be syslog message for same ?

File: tests/everflow/everflow_test_utilities.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        session_dst_ip = "2.2.2.2"
+        session_ttl = "1"
+        session_dscp = "8"
+
+        if "mellanox" == duthost.facts["asic_type"]:
+            session_gre = 0x8949
+        elif "barefoot" == duthost.facts["asic_type"]:
+            session_gre = 0x22EB
+        else:
+            session_gre = 0x6558
```
Comment (yxieca) [inline tests/everflow/everflow_test_utilities.py:235]: Probably should define these codes in a more common place so that other tests could share. (Not a change request for this PR).

**Comment by yxieca** [comment]
retest vsimage please

File: tests/everflow/everflow_test_utilities.py (hunk 6 above)
Comment (daall) [reply tests/everflow/everflow_test_utilities.py:235]: Ack, I'm going to go through and refactor the existing test cases so that all the shared dependencies are in everflow_utilities so I will include this in that effort.

File: tests/everflow/templates/acl_rule_v6.json.j2 (hunk 1 above)
Comment (daall) [reply tests/everflow/templates/acl_rule_v6.json.j2:370]: Nope, no error should occur. The ASIC should be able to handle arbitrary protocol numbers supplied by the user. I think it might be good to include some other random protocol numbers as well just to be sure.

File: tests/everflow/everflow_test_utilities.py [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+        test_ingress_mirror_on_egress_acl = False
+        test_egress_mirror_on_egress_acl = False
+        test_egress_mirror_on_ingress_acl = False
+    else:
+        test_ingress_mirror_on_ingress_acl = "MIRROR_INGRESS_ACTION" in switch_capabilities["ACL_ACTIONS|INGRESS"]
+        test_ingress_mirror_on_egress_acl = "MIRROR_INGRESS_ACTION" in switch_capabilities["ACL_ACTIONS|EGRESS"]
+        test_egress_mirror_on_egress_acl = "MIRROR_EGRESS_ACTION" in switch_capabilities["ACL_ACTIONS|EGRESS"]
+        test_egress_mirror_on_ingress_acl = "MIRROR_EGRESS_ACTION" in switch_capabilities["ACL_ACTIONS|INGRESS"]
+
+    def get_port_info(in_port_list, out_port_list, out_port_ptf_id_list, out_port_lag_name):
```
Comment (lguohan) [inline tests/everflow/everflow_test_utilities.py:66]: better comment this function, what kind of port info do you get?
